File: extensions/sdk-python/nexo_plugin_sdk/broker.py
"""Child-side broker handle.

Plugin authors call ``broker.publish(topic, event)`` to emit
notifications back to the daemon. Topics MUST appear on the
manifest's ``[[plugin.channels.register]]`` allowlist or the
daemon drops the message with a warn log (defense in depth on
the host side).
"""

import asyncio
import sys
from typing import TextIO

from . import wire
from .events import Event


class BrokerSender:
    """Write-only handle to the daemon's broker.

    Wraps an injected line-writer (the captured *original* stdout,
    so blessed JSON-RPC frames bypass the stdout guard) behind an
    async lock so concurrent handler tasks do not interleave
    half-written frames.
    """

    def __init__(self, write_lock: asyncio.Lock, writer: TextIO | None = None) -> None:
        self._lock = write_lock
        # Default to bare ``sys.stdout`` for callers (mainly tests)
        # that construct a BrokerSender without going through the
        # adapter. The adapter always injects the captured original.
        self._writer: TextIO = writer if writer is not None else sys.stdout

    async def publish(self, topic: str, event: Event) -> None:
        line = wire.serialize_frame(
            wire.build_notification(
                "broker.publish", {"topic": topic, "event": event.to_json()}
            )
        )
        async with self._lock:
            self._writer.write(line)
            self._writer.flush()

File: extensions/sdk-typescript/src/broker.ts
/**
 * Phase 31.5 — child-side broker handle.
 *
 * Plugin authors call `broker.publish(topic, event)` to emit
 * notifications back to the daemon. Topics MUST appear on the
 * manifest's `[[plugin.channels.register]]` allowlist; the host
 * drops disallowed topics with a warn log (defense in depth).
 *
 * Concurrency: a Promise-chain write lock serializes outbound
 * frames so concurrent handler tasks never interleave half-written
 * JSON lines on stdout. Callers `await broker.publish(...)`; the
 * returned Promise resolves once the byte sequence has been flushed
 * to the parent's stdin.
 */

import { Event } from "./events.js";
import { JSONRPC_VERSION, serializeFrame } from "./wire.js";

export type LineWriter = (line: string) => void;

export class BrokerSender {
  /** Bounded by the JS event loop's microtask queue — adds zero
   * scheduling overhead per call. Each `publish` chains
   * `writeChain = writeChain.then(doWrite)`, so the second caller
   * waits until the first completes synchronously inside the
   * thenable, preserving FIFO order without async semaphores.
   */
  private writeChain: Promise<void> = Promise.resolve();
  private readonly write: LineWriter;

  constructor(write: LineWriter) {
    this.write = write;
  }

  async publish(topic: string, event: Event): Promise<void> {
    const next = this.writeChain.then(() => this.doWrite(topic, event));
    this.writeChain = next.catch(() => {
      // Surface the error to the awaiting caller via the next
      // assignment, but reset the chain so a single failed write
      // does not poison every subsequent publish.
    });
    return next;
  }

  private doWrite(topic: string, event: Event): void {
    const frame = {
      jsonrpc: JSONRPC_VERSION as typeof JSONRPC_VERSION,
      method: "broker.publish",
      params: { topic, event: Event.toJson(event) },
    };
    this.write(serializeFrame(frame));
  }
}
